scripts/jenkins/ardana/gerrit/build_test_package.py

```diff
@@ -161,33 +171,49 @@ def cleanup(testproject):
         shutil.rmtree('./%s' % testproject)
 
 
-def main():
-    change_ids = os.environ['gerrit_change_ids'].split(',')
-    develproject = os.environ['develproject']
-    homeproject = os.environ['homeproject']
-    repository = os.environ['repository']
+def main(change_ids, develproject, homeproject, repository):
     testproject = test_project_name(change_ids, homeproject)
-    changes = [GerritChange(id) for id in change_ids]
     cleanup(testproject)
-    testproject = create_test_project(develproject, testproject, repository)
-
-    # Create the test packages to be built
-    # NOTE(jhesketh): If necessary this could be done in a threadpool
-    for change in changes:
+    create_test_project(develproject, testproject, repository)
+
+    # Changes is dict of 'gerrit_id': GerritChange.
+    changes = {}
+    # Keep track of the list of projects to finally build
+    projects = set()
+    # Grab the changes
+    for id in change_ids:
+        if id in changes:
+            # Duplicate dependency, skipping..
+            continue
+        c = GerritChange(id)
+        changes[c.change_id] = c
+        projects.add(c.project)
+        dependent_changes = c.get_depends_on_changes()
+        change_ids.extend(dependent_changes)
+
+    for change in changes.values():
+        # This merges the particular gerrit change into the workspace
         change.prep_workspace()
-        create_test_package(change, develproject, testproject)
+
+    for project in projects:
+        create_test_package(project, develproject, testproject)
 
     # Check all packages are built
     # NOTE(jhesketh): this could be optimised to check packages in parallel,
     # however the worst case scenario at the moment is
     # "time for longest package" + "time for num_of_package checks" which isn't
     # too much more than the minimum
     # ("time for longest package" + "time for one check")
-    for change in changes:
-        wait_for_build(change,  testproject)
+    for project in projects:
+        wait_for_build(project,  testproject)
 
     cleanup(testproject)
 
 
 if __name__ == '__main__':
-    main()
+    change_ids = os.environ['gerrit_change_ids'].split(',')
+    develproject = os.environ['develproject']
+    homeproject = os.environ['homeproject']
+    repository = os.environ['repository']
+
+    main(change_ids, develproject, homeproject, repository)
```

Comment:
Please don't put anything besides one function call into ``__main__``. E.g. by renaming main as it is now to build_test_packages and moving the above to a new main.